### PR TITLE
Minor clarifications

### PR DESCRIPTION
**Pull request by toote**
After some comments in [issue 372 in the plugin linter](https://github.com/buildkite-plugins/buildkite-plugin-linter/issues/372) I believe that some clarifications are necessary on the documentation to write plugins with regards to the `builkite-plugin` suffix and vendored plugins.

### CODE REVIEW

File: pages/plugins/writing.md.erb [hunk 1]
```diff
@@ -46,10 +49,10 @@ configuration:
 
 The `configuration` property defines the validation rules for the plugin configuration using the [JSON Schema](https://json-schema.org) format. The plugin in this tutorial has a single `pattern` property, of type `string`.
 
-Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is the GitHub repository slug, not the name defined in `plugin.yml`. In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
+Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but an uppercase version of the plugin's ID. In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:52]: Is the _plugin ID_ defined or referenced anywhere else? Does it mean a lowercase version of the plugin name, with any whitespace replaced with dashes?

File: pages/plugins/writing.md.erb [hunk 2]
```diff
@@ -46,10 +49,10 @@ configuration:
...
 
->🚧 Plugin properties and git
-> Note that if you <a href="/docs/plugins/using#plugin-sources">reference a plugin</a> using a full URL and <code>.git</code> extension, you need to use a slightly different syntax to get access to the configuration properties. For example, to get the value of <code>pattern</code> in <code>https://github.com/my-org/my-plugin.git#v1.0.0</code> use  <code>BUILDKITE_PLUGIN_MY_PLUGIN_GIT_PATTERN</code>.
+>🚧 Variables and non-standard git plugins
+> Note that if you <a href="/docs/plugins/using#plugin-sources">reference a plugin</a> with a full URL ending in <code>.git</code> and that plugin's name does not end with `buildkite-plugin`, variable names will include `_GIT` as if it were part of the plugin name. For example, the value of the configuration <code>pattern</code> in <code>https://github.com/my-org/my-plugin.git#v1.0.0</code> will be available as <code>BUILDKITE_PLUGIN_MY_PLUGIN_GIT_PATTERN</code>.
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:55]: Just a couple of nitpicks.

```suggestion
> Note that if you <a href="/docs/plugins/using#plugin-sources">reference a plugin</a> with a full URL ending in <code>.git</code> and that plugin's name does not end with `-buildkite-plugin`, variable names will include `_GIT` as part of the plugin name. For example, the value of the configuration <code>pattern</code> in <code>https://github.com/my-org/my-plugin.git#v1.0.0</code> will be available as <code>BUILDKITE_PLUGIN_MY_PLUGIN_GIT_PATTERN</code>.
```

File: pages/plugins/writing.md.erb [hunk 3]
```diff
@@ -46,10 +49,10 @@ configuration:
...
+Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but an uppercase version of the plugin's ID. In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
 
->🚧 Plugin properties and git
-> Note that if you <a href="/docs/plugins/using#plugin-sources">reference a plugin</a> using a full URL and <code>.git</code> extension, you need to use a slightly different syntax to get access to the configuration properties. For example, to get the value of <code>pattern</code> in <code>https://github.com/my-org/my-plugin.git#v1.0.0</code> use  <code>BUILDKITE_PLUGIN_MY_PLUGIN_GIT_PATTERN</code>.
+>🚧 Variables and non-standard git plugins
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:54]: The rest of the page frames these ideas in terms of _properties_ rather than _variables_, so I think it would be good to stick to that. I think it could be more descriptive. Thinking:

```suggestion
>🚧 Accessing plugin properties with Git URLs
```

(_Git_ should also be capitalized)

File: pages/plugins/writing.md.erb [hunk 4]
```diff
@@ -94,7 +97,20 @@ You can run the plugin linter with the following Docker command:
 docker run -it --rm -v "$PWD:/plugin:ro" buildkite/plugin-linter --id a-github-user/file-counter
 ```
 
-To make it easier to run this command, add it to the `docker-compose.yml` file:
+### Plugin linter plugin
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:100]: I don't think we need the repetition since it's clear from the plugin name.

```suggestion
### Plugin linter
```

File: pages/plugins/writing.md.erb [hunk 5]
```diff
@@ -94,7 +97,20 @@ You can run the plugin linter with the following Docker command:
...
+
+If your plugin has a Buildkite pipeline, you can add a step to lint it using the corresponding plugin:
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:102]: It would be good to be more specific here:

```suggestion
If your plugin has a Buildkite pipeline, you can add a step to lint it using the Plugin linter:
```

**Review by mbelton-buildkite — COMMENTED** [review]
Thanks for putting this together! Sorry it's taken so long to move forward. I've left some comments for how I think we can adjust it before merging.

File: pages/plugins/writing.md.erb (hunk 3 above)
Comment (toote) [reply pages/plugins/writing.md.erb:54]: How about the follwing to avoid a potential confusion of Git URLs being the values of the plugin properties?

```suggestion
>🚧 Accessing properties on plugins referenced with Git URLs
```

File: pages/plugins/writing.md.erb (hunk 4 above)
Comment (toote) [reply pages/plugins/writing.md.erb:100]: Not really... this is a buildkite plugin that runs the _linter for buildkite plugins_ so the plugin is not a repetition but an actual description that is necessary to differentiate between the linter itself and the plugin that runs the linter

Comment (toote) [reply pages/plugins/writing.md.erb:100]: Maybe a better wording would be:
```suggestion
### Buildkite Plugin to run the plugin linter
```

File: pages/plugins/writing.md.erb (hunk 5 above)
Comment (toote) [reply pages/plugins/writing.md.erb:102]: Same here... we are talking about the buildkite plugin that runs the plugin linter (not the plugin linter itself). Yeah, I know...

File: pages/plugins/writing.md.erb [hunk 6]
```diff
@@ -94,7 +97,20 @@ You can run the plugin linter with the following Docker command:
...
+
+```yml
+  - label: ":shell: Lint"
+    plugins:
+      plugin-linter#v3.0.0:
+        id: a-gihub-user/file-counter
+```
+{: codeblock-file=".buildkite/pipeline.yml"}
+
+### Docker compose
```
[the remark this answers is not in the document]
Comment (toote) [reply pages/plugins/writing.md.erb:112]: Indeed... the two sub headings are two ways to make it easier to run the docker command by hand: using a buildkite plugin or docker compose

File: pages/plugins/writing.md.erb (hunk 1 above)
Comment (toote) [reply pages/plugins/writing.md.erb:52]: That is a good point... I don't think that there is a full documentation of how the ID is created. Environment variables are [created by the agent](https://github.com/buildkite/agent/blob/ec8675e062f4c9e83e772a827305b66331f1ba7a/agent/plugin/plugin.go#L249). Following the code, it would appear that - at this time at least - the logic is as follows:

* take [the plugin name](https://github.com/buildkite/agent/blob/ec8675e062f4c9e83e772a827305b66331f1ba7a/agent/plugin/plugin.go#L139):
  - from the plugin's definition, take whatever is after the last slash (`/`)
  - make it all lowercase
  - replace every whitespace in it with a single space
  - replace every character that is not a number or letter with a dash `-`
  - remove the `-buildkite-plugin(-git)` suffix
* convert it to a [variable name](https://github.com/buildkite/agent/blob/ec8675e062f4c9e83e772a827305b66331f1ba7a/agent/plugin/plugin.go#L207):
  - make it all uppercase
  - replace every whitespace in it with a single space
  - dashes and spaces are replaced with underscores
  - if there are contiguous underscores, they are replaced with a single one

So, basically, it is mostly what you said... except it is not the plugin name, but how it is referenced in the pipeline itself.

I used `plugin id` because that is how the linter defines it, but I don't think that it makes sense to use it.

How about the following?

```suggestion
Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but a variable-friendly (only letters, numbers and underscores) version of the repository or folder name in uppercase. In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
```

File: pages/plugins/writing.md.erb [hunk 7]
```diff
@@ -238,3 +254,20 @@ When writing plugins, there are two patterns you can choose from:
...
+```yml
+steps:
+  - command: ls
+    plugins:
+      - ./relative/path/to/plugin:
+          pattern: '*.md'
+```
+{: codeblock-file="pipeline.yml"}
+
+Make sure to review [the documentation about job lifecycle hooks](docs/agent/v3/hooks#job-lifecycle-hooks) to understand when vendored plugins are run. Not only they are normally run after non-vendored plugins, but also some hooks are not available.
```
[the remark this answers is not in the document]
Comment (toote) [reply pages/plugins/writing.md.erb]: Looks good except for the link so I committed it myself.

File: pages/plugins/writing.md.erb [hunk 8]
```diff
@@ -19,14 +19,17 @@ steps:
...
+The most common kind of Buildkite plugin is a Git repository, with a descriptive name ending in `-buildkite-plugin`. This suffix is required: Buildkite automatically adds `-buildkite-plugin` to the plugin name specified in the `pipeline.yml`. Let's create a new Git repository following these naming conventions:
 
 ```shell
 mkdir file-counter-buildkite-plugin
 cd file-counter-buildkite-plugin
 git init
 ```
 
+> 🚧 The `buildkite-plugin` suffix
+> Not including `-buildkite-plugin` as the end of your plugin's repository is possible but may cause you unexpected issues. It will also prevent other members of the community from easily finding and using the plugin.
```
[the remark this answers is not in the document]
Comment (toote) [reply pages/plugins/writing.md.erb]: I reworded this notice and the previous one to make the impact more clear

File: pages/plugins/writing.md.erb (hunk 1 above)
Comment (toote) [reply pages/plugins/writing.md.erb:52]: Another alternative would be:

```suggestion
Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but the repository or folder name compatible with Bash environment variables (in uppercase and only letters, numbers and underscores). In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
```

File: pages/plugins/writing.md.erb [hunk 9]
```diff
@@ -19,14 +19,17 @@ steps:
...
-Every Buildkite plugin is a Git repository, ending in `-buildkite-plugin`. This suffix is required: Buildkite automatically adds `-buildkite-plugin` to the plugin name specified in the `pipeline.yml`. Let's create a new Git repository following these naming conventions:
+The most common kind of Buildkite plugin is a Git repository, with a descriptive name ending in `-buildkite-plugin`. This suffix is required to allow using the `user/plugin-name` syntax in pipelines. Let's create a new Git repository following these naming conventions:
 
 ```shell
 mkdir file-counter-buildkite-plugin
 cd file-counter-buildkite-plugin
 git init
 ```
 
+> 🚧 The `-buildkite-plugin` suffix
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:30]: I think this is more of a note/tip than a warning.

```suggestion
> 📘 The `-buildkite-plugin` suffix
```

File: pages/plugins/writing.md.erb [hunk 10]
```diff
@@ -19,14 +19,17 @@ steps:
...
+> Not including the `-buildkite-plugin` suffix in the repository name not only will prevent you from using the `user/plugin-name` syntax in the pipeline (you will need to reference the plugin with a full URL instead) but also makes it **more difficult for community members to find and use the plugin**. And even when you know it will be used through a full URL, only in your organization, or it is a vendored plugin the suffix should help identify the purpose of its contents to anyone looking at it.
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:31]: I think this reads more clearly and breaks up the paragraph:

```suggestion
> We recommend using the `-buildkite-plugin` suffix in the repository name because:
<ul>
  <li>You can reference the plugin in pipelines using the <code>user/plugin-name</code> syntax rather than the full URL.</li>
  <li>It makes it easier for community members to find and use the plugin if you make it public.</li>
  <li>It communicates the purpose of the code.</li>
</ul>
```

File: pages/plugins/writing.md.erb [hunk 11]
```diff
@@ -46,10 +49,10 @@ configuration:
...
+Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but the repository or folder name compatible with Bash environment variables (in uppercase and only letters, numbers and underscores). In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
 
->🚧 Plugin properties and git
-> Note that if you <a href="/docs/plugins/using#plugin-sources">reference a plugin</a> using a full URL and <code>.git</code> extension, you need to use a slightly different syntax to get access to the configuration properties. For example, to get the value of <code>pattern</code> in <code>https://github.com/my-org/my-plugin.git#v1.0.0</code> use  <code>BUILDKITE_PLUGIN_MY_PLUGIN_GIT_PATTERN</code>.
+>🚧 Accessing properties on plugins referenced with Git URLs
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:54]: I think this is also more a note than a warning. We hope most people aren't running into this because they've followed the guidance above about including the suffix.

```suggestion
>📘 Accessing properties on plugins referenced with Git URLs
```

File: pages/plugins/writing.md.erb [hunk 12]
```diff
@@ -94,7 +97,20 @@ You can run the plugin linter with the following Docker command:
 docker run -it --rm -v "$PWD:/plugin:ro" buildkite/plugin-linter --id a-github-user/file-counter
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:97]: I think we should replace all of `## Step 3: Validate the plugin.yml` with:

````
## Step 3: Validate the plugin.yml

The [Buildkite Plugin Linter](https://github.com/buildkite-plugins/buildkite-plugin-linter) is an app that helps ensure your plugin is up-to-date and has all the files required to list it in the plugins directory. The app is available as a Docker image you can run on the command line, with a dedicated plugin, or with Docker Compose. We recommend you start by running the linter on the command line, and then include the dedicated plugin in the pipeline for your plugin.

### Run on the command line

You can run the plugin linter with the following Docker command:

```shell
docker run -it --rm -v "$PWD:/plugin:ro" buildkite/plugin-linter --id a-github-user/file-counter
```

### Run with the dedicated plugin

If your plugin has a Buildkite pipeline, you can add a step to lint it using the corresponding plugin:

```yml
  - label: ":shell: Lint"
    plugins:
      plugin-linter#v3.0.0:
        id: a-gihub-user/file-counter
```
{: codeblock-file=".buildkite/pipeline.yml"}

### Run with Docker Compose

If you want to run the linter using Docker Compose, you can add the following to a `docker-compose.yml` file:

```yml
services:
  lint:
    image: buildkite/plugin-linter
    command: ['--id', 'a-github-user/file-counter']
    volumes:
      - ".:/plugin:ro"
```
{: codeblock-file="docker-compose.yml"} 

You can then run the tests using the following command:

```shell
docker-compose run --rm lint
```
````

What do you think?

**Review by mbelton-buildkite — COMMENTED** [review]
I tried to push a commit but was told I didn't have write access. These are the changes I think we should make.

File: pages/plugins/writing.md.erb (hunk 10 above)
Comment (toote) [reply pages/plugins/writing.md.erb]: I didn't quite like the mix of HTML... but could not find a work around it :(

File: pages/plugins/writing.md.erb (hunk 12 above)
Comment (toote) [reply pages/plugins/writing.md.erb:104]: Used as-is, except for the main heading that I changed to `Validate the plugin` (as the linter checks more things than just the `plugin.yml` file)

File: pages/plugins/writing.md.erb [hunk 13]
```diff
@@ -46,10 +54,10 @@ configuration:
 
 The `configuration` property defines the validation rules for the plugin configuration using the [JSON Schema](https://json-schema.org) format. The plugin in this tutorial has a single `pattern` property, of type `string`.
 
-Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is the GitHub repository slug, not the name defined in `plugin.yml`. In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
+Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but the repository or folder name compatible with Bash environment variables (in uppercase and only letters, numbers and underscores). In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:57]: One last nitpick: Use a serial comma.
```suggestion
Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but the repository or folder name compatible with Bash environment variables (in uppercase and only letters, numbers, and underscores). In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
```

**Review by mbelton-buildkite — APPROVED** [review]
Looks good to me! Great work on this 🙌🏻

I'm just running the tests and then I'll merge for you (I don't think you have all the permissions to kick off a deploy).

File: pages/plugins/writing.md.erb [hunk 14]
```diff
@@ -238,3 +262,18 @@ When writing plugins, there are two patterns you can choose from:
...
+```yml
+steps:
+  - command: ls
+    plugins:
+      - ./relative/path/to/plugin:
+          pattern: '*.md'
+```
+{: codeblock-file="pipeline.yml"}
+
+Vendored plugins run after non-vendored plugins and don't have access to all the same hooks. See [the documentation about job lifecycle hooks](docs/agent/v3/hooks#job-lifecycle-hooks) to learn more.
```
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:279]: Fix broken link
```suggestion
Vendored plugins run after non-vendored plugins and don't have access to all the same hooks. See [the documentation about job lifecycle hooks](/docs/agent/v3/hooks#job-lifecycle-hooks) to learn more.
```
